# Formula 1 Driver Performance Analysis (2023-2024)

The goal of this notebook is to analyze the changes in lap times and overall performance of each driver from the 2023 to the 2024 Formula 1 Season.

### Table of Contents
- [Load Data](#Load-Data)
- [Data Cleaning](#Data-Cleaning)
- [Data Preprocessing](#Data-Preprocessing)
- [Data Visualization](#Data-Visualization)

### Import Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import os
import seaborn as sns

### Load Data

First, we need to load the necessary data we're going to be using into memory.

Declare path where data is stored

In [2]:
PATH = '../data/'

Load drivers dataframe

In [3]:
df_drivers = pd.read_csv(PATH+'drivers.csv')
df_drivers.shape

(859, 9)

In [4]:
df_drivers.head()

driverId   driverRef number code  forename     surname         dob  \
0         1    hamilton     44  HAM     Lewis    Hamilton  1985-01-07   
1         2    heidfeld     \N  HEI      Nick    Heidfeld  1977-05-10   
2         3     rosberg      6  ROS      Nico     Rosberg  1985-06-27   
3         4      alonso     14  ALO  Fernando      Alonso  1981-07-29   
4         5  kovalainen     \N  KOV    Heikki  Kovalainen  1981-10-19   

  nationality                                             url  
0     British     http://en.wikipedia.org/wiki/Lewis_Hamilton  
1      German      http://en.wikipedia.org/wiki/Nick_Heidfeld  
2      German       http://en.wikipedia.org/wiki/Nico_Rosberg  
3     Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso  
4     Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen

Load lap times dataframe

In [5]:
df_lap_times = pd.read_csv(PATH+'lap_times.csv')
df_lap_times.shape

(575029, 6)

In [6]:
df_lap_times.head()

raceId  driverId  lap  position      time  milliseconds
0     841        20    1         1  1:38.109         98109
1     841        20    2         1  1:33.006         93006
2     841        20    3         1  1:32.713         92713
3     841        20    4         1  1:32.803         92803
4     841        20    5         1  1:32.342         92342

Load races dataframe

In [7]:
df_races = pd.read_csv(PATH+'races.csv')
df_races.shape

(1125, 18)

In [8]:
df_races.head()

raceId  year  round  circuitId                   name        date  \
0       1  2009      1          1  Australian Grand Prix  2009-03-29   
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05   
2       3  2009      3         17     Chinese Grand Prix  2009-04-19   
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26   
4       5  2009      5          4     Spanish Grand Prix  2009-05-10   

       time                                                url fp1_date  \
0  06:00:00  http://en.wikipedia.org/wiki/2009_Australian_G...       \N   
1  09:00:00  http://en.wikipedia.org/wiki/2009_Malaysian_Gr...       \N   
2  07:00:00  http://en.wikipedia.org/wiki/2009_Chinese_Gran...       \N   
3  12:00:00  http://en.wikipedia.org/wiki/2009_Bahrain_Gran...       \N   
4  12:00:00  http://en.wikipedia.org/wiki/2009_Spanish_Gran...       \N   

  fp1_time fp2_date fp2_time fp3_date fp3_time quali_date quali_time  \
0       \N       \N       \N       \N       \N         \N         \N   
1       \N       \N       \N       \N       \N         \N         \N   
2       \N       \N       \N       \N       \N         \N         \N   
3       \N       \N       \N       \N       \N         \N         \N   
4       \N       \N       \N       \N       \N         \N         \N   

  sprint_date sprint_time  
0          \N          \N  
1          \N          \N  
2          \N          \N  
3          \N          \N  
4          \N          \N

Load results dataframe

In [9]:
df_results = pd.read_csv(PATH+'results.csv')
df_results.shape

(26519, 18)

In [10]:
df_results.head()

resultId  raceId  driverId  constructorId number  grid position  \
0         1      18         1              1     22     1        1   
1         2      18         2              2      3     5        2   
2         3      18         3              3      7     7        3   
3         4      18         4              4      5    11        4   
4         5      18         5              1     23     3        5   

  positionText  positionOrder  points  laps         time milliseconds  \
0            1              1    10.0    58  1:34:50.616      5690616   
1            2              2     8.0    58       +5.478      5696094   
2            3              3     6.0    58       +8.163      5698779   
3            4              4     5.0    58      +17.181      5707797   
4            5              5     4.0    58      +18.014      5708630   

  fastestLap rank fastestLapTime fastestLapSpeed  statusId  
0         39    2       1:27.452         218.300         1  
1         41    3       1:27.739         217.586         1  
2         41    5       1:28.090         216.719         1  
3         58    7       1:28.603         215.464         1  
4         43    1       1:27.418         218.385         1

### Data Successfully Loaded:
```
data
└───drivers.csv          -> df_drivers
└───lap_times.csv        -> df_lap_times
└───races.csv            -> df_races
└───results.csv          -> df_results
```

### Data Cleaning

Let's make sure we don't run into any errors during analysis by cleaning up any null values from the data.

Check for any missing data or null values

In [11]:
df_drivers.isnull().any()

driverId       False
driverRef      False
number         False
code           False
forename       False
surname        False
dob            False
nationality    False
url            False
dtype: bool

In [12]:
df_lap_times.isnull().any()

raceId          False
driverId        False
lap             False
position        False
time            False
milliseconds    False
dtype: bool

In [13]:
df_races.isnull().any()

raceId         False
year           False
round          False
circuitId      False
name           False
date           False
time           False
url            False
fp1_date       False
fp1_time       False
fp2_date       False
fp2_time       False
fp3_date       False
fp3_time       False
quali_date     False
quali_time     False
sprint_date    False
sprint_time    False
dtype: bool

In [14]:
df_results.isnull().any()

resultId           False
raceId             False
driverId           False
constructorId      False
number             False
grid               False
position           False
positionText       False
positionOrder      False
points             False
laps               False
time               False
milliseconds       False
fastestLap         False
rank               False
fastestLapTime     False
fastestLapSpeed    False
statusId           False
dtype: bool

Luckily, none of these datasets have any missing values so we can go straight to preprocessing.

### Data Preprocessing

#### Initial Look at Data

##### Let's start by looking at df_drivers

In [15]:
df_drivers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 859 entries, 0 to 858
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   driverId     859 non-null    int64 
 1   driverRef    859 non-null    object
 2   number       859 non-null    object
 3   code         859 non-null    object
 4   forename     859 non-null    object
 5   surname      859 non-null    object
 6   dob          859 non-null    object
 7   nationality  859 non-null    object
 8   url          859 non-null    object
dtypes: int64(1), object(8)
memory usage: 60.5+ KB


In [16]:
df_drivers.head()

driverId   driverRef number code  forename     surname         dob  \
0         1    hamilton     44  HAM     Lewis    Hamilton  1985-01-07   
1         2    heidfeld     \N  HEI      Nick    Heidfeld  1977-05-10   
2         3     rosberg      6  ROS      Nico     Rosberg  1985-06-27   
3         4      alonso     14  ALO  Fernando      Alonso  1981-07-29   
4         5  kovalainen     \N  KOV    Heikki  Kovalainen  1981-10-19   

  nationality                                             url  
0     British     http://en.wikipedia.org/wiki/Lewis_Hamilton  
1      German      http://en.wikipedia.org/wiki/Nick_Heidfeld  
2      German       http://en.wikipedia.org/wiki/Nico_Rosberg  
3     Spanish    http://en.wikipedia.org/wiki/Fernando_Alonso  
4     Finnish  http://en.wikipedia.org/wiki/Heikki_Kovalainen

This database appears to only be useful as a reference for which driver driverId refers to. We can drop most of the columns as for our analysis forename, surname, dob, nationality, and their wikipedia url (why?) will not be useful to us at all. I'll leave in number and code because they may be useful for labeling purposes down the line. Surname might have been useful, but since it looks like driverRef is just the driver's surname in lowercase we'll be dropping it for sure.

In [17]:
df_drivers.drop(labels=['forename', 'surname', 'dob', 'nationality', 'url'], axis='columns', inplace=True)

In [18]:
df_drivers.head()

driverId   driverRef number code
0         1    hamilton     44  HAM
1         2    heidfeld     \N  HEI
2         3     rosberg      6  ROS
3         4      alonso     14  ALO
4         5  kovalainen     \N  KOV

##### Let's look at df_lap_times

In [19]:
df_lap_times.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 575029 entries, 0 to 575028
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   raceId        575029 non-null  int64 
 1   driverId      575029 non-null  int64 
 2   lap           575029 non-null  int64 
 3   position      575029 non-null  int64 
 4   time          575029 non-null  object
 5   milliseconds  575029 non-null  int64 
dtypes: int64(5), object(1)
memory usage: 26.3+ MB


In [20]:
df_lap_times.describe()

raceId       driverId            lap       position  \
count  575029.000000  575029.000000  575029.000000  575029.000000   
mean      587.413494     315.316934      30.021122       9.659650   
std       431.352273     384.339100      18.417608       5.531594   
min         1.000000       1.000000       1.000000       1.000000   
25%       137.000000      15.000000      14.000000       5.000000   
50%       855.000000      44.000000      29.000000       9.000000   
75%       993.000000     820.000000      44.000000      14.000000   
max      1131.000000     860.000000      87.000000      24.000000   

       milliseconds  
count  5.750290e+05  
mean   9.580980e+04  
std    7.692607e+04  
min    5.540400e+04  
25%    8.196200e+04  
50%    9.062700e+04  
75%    1.019710e+05  
max    7.507547e+06

In [21]:
df_lap_times.head()

raceId  driverId  lap  position      time  milliseconds
0     841        20    1         1  1:38.109         98109
1     841        20    2         1  1:33.006         93006
2     841        20    3         1  1:32.713         92713
3     841        20    4         1  1:32.803         92803
4     841        20    5         1  1:32.342         92342

It would appear that this is the largest database for good reason. It looks like it contains every single lap time driven for every single driver in every single race. With there being 1131 unique raceIds there over 1000 races worth of lap times for each race with each race including 20-24 drivers and 305km worth of laps on the track. (Every F1 race is just the number of laps it takes to reach 305km driven).

All of these columns will be useful for data analysis.

##### Let's look at df_races

In [22]:
df_races.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1125 entries, 0 to 1124
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   raceId       1125 non-null   int64 
 1   year         1125 non-null   int64 
 2   round        1125 non-null   int64 
 3   circuitId    1125 non-null   int64 
 4   name         1125 non-null   object
 5   date         1125 non-null   object
 6   time         1125 non-null   object
 7   url          1125 non-null   object
 8   fp1_date     1125 non-null   object
 9   fp1_time     1125 non-null   object
 10  fp2_date     1125 non-null   object
 11  fp2_time     1125 non-null   object
 12  fp3_date     1125 non-null   object
 13  fp3_time     1125 non-null   object
 14  quali_date   1125 non-null   object
 15  quali_time   1125 non-null   object
 16  sprint_date  1125 non-null   object
 17  sprint_time  1125 non-null   object
dtypes: int64(4), object(14)
memory usage: 158.3+ KB


In [23]:
df_races.describe()

raceId         year        round    circuitId
count  1125.000000  1125.000000  1125.000000  1125.000000
mean    565.710222  1992.703111     8.579556    23.889778
std     328.813817    20.603848     5.159910    19.633527
min       1.000000  1950.000000     1.000000     1.000000
25%     282.000000  1977.000000     4.000000     9.000000
50%     563.000000  1994.000000     8.000000    18.000000
75%     845.000000  2011.000000    13.000000    34.000000
max    1144.000000  2024.000000    24.000000    80.000000

In [24]:
df_races.head()

raceId  year  round  circuitId                   name        date  \
0       1  2009      1          1  Australian Grand Prix  2009-03-29   
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05   
2       3  2009      3         17     Chinese Grand Prix  2009-04-19   
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26   
4       5  2009      5          4     Spanish Grand Prix  2009-05-10   

       time                                                url fp1_date  \
0  06:00:00  http://en.wikipedia.org/wiki/2009_Australian_G...       \N   
1  09:00:00  http://en.wikipedia.org/wiki/2009_Malaysian_Gr...       \N   
2  07:00:00  http://en.wikipedia.org/wiki/2009_Chinese_Gran...       \N   
3  12:00:00  http://en.wikipedia.org/wiki/2009_Bahrain_Gran...       \N   
4  12:00:00  http://en.wikipedia.org/wiki/2009_Spanish_Gran...       \N   

  fp1_time fp2_date fp2_time fp3_date fp3_time quali_date quali_time  \
0       \N       \N       \N       \N       \N         \N         \N   
1       \N       \N       \N       \N       \N         \N         \N   
2       \N       \N       \N       \N       \N         \N         \N   
3       \N       \N       \N       \N       \N         \N         \N   
4       \N       \N       \N       \N       \N         \N         \N   

  sprint_date sprint_time  
0          \N          \N  
1          \N          \N  
2          \N          \N  
3          \N          \N  
4          \N          \N

This dataset looks like it's similar to the drivers reference dataset but for the formula 1 tracks themselves. There seems to be a lot of messy columns in here and as this is just an analysis of lap times, we can start by removing most of the non-useful columns.

In [25]:
df_races.drop(
    labels=['url', 
            'fp1_date', 'fp1_time', 
            'fp2_date', 'fp2_time', 
            'fp3_date', 'fp3_time', 
            'quali_date', 'quali_time', 
            'sprint_date', 'sprint_time'],
    axis='columns',
    inplace=True)

In [26]:
df_races.head()

raceId  year  round  circuitId                   name        date      time
0       1  2009      1          1  Australian Grand Prix  2009-03-29  06:00:00
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05  09:00:00
2       3  2009      3         17     Chinese Grand Prix  2009-04-19  07:00:00
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26  12:00:00
4       5  2009      5          4     Spanish Grand Prix  2009-05-10  12:00:00

##### Let's look at df_results

In [27]:
df_results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26519 entries, 0 to 26518
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         26519 non-null  int64  
 1   raceId           26519 non-null  int64  
 2   driverId         26519 non-null  int64  
 3   constructorId    26519 non-null  int64  
 4   number           26519 non-null  object 
 5   grid             26519 non-null  int64  
 6   position         26519 non-null  object 
 7   positionText     26519 non-null  object 
 8   positionOrder    26519 non-null  int64  
 9   points           26519 non-null  float64
 10  laps             26519 non-null  int64  
 11  time             26519 non-null  object 
 12  milliseconds     26519 non-null  object 
 13  fastestLap       26519 non-null  object 
 14  rank             26519 non-null  object 
 15  fastestLapTime   26519 non-null  object 
 16  fastestLapSpeed  26519 non-null  object 
 17  statusId    

In [28]:
df_results.describe()

resultId        raceId      driverId  constructorId          grid  \
count  26519.000000  26519.000000  26519.000000   26519.000000  26519.000000   
mean   13260.940986    546.376560    274.357291      49.801161     11.145820   
std     7656.813206    309.642244    279.275606      61.091426      7.213453   
min        1.000000      1.000000      1.000000       1.000000      0.000000   
25%     6630.500000    298.000000     57.000000       6.000000      5.000000   
50%    13260.000000    527.000000    170.000000      25.000000     11.000000   
75%    19889.500000    803.000000    385.000000      60.000000     17.000000   
max    26524.000000   1132.000000    860.000000     215.000000     34.000000   

       positionOrder        points          laps      statusId  
count   26519.000000  26519.000000  26519.000000  26519.000000  
mean       12.814812      1.959578     46.228251     17.317056  
std         7.677869      4.306475     29.577860     26.081700  
min         1.000000      0.000000      0.000000      1.000000  
25%         6.000000      0.000000     23.000000      1.000000  
50%        12.000000      0.000000     53.000000     10.000000  
75%        18.000000      2.000000     66.000000     14.000000  
max        39.000000     50.000000    200.000000    141.000000

In [29]:
df_results.head()

resultId  raceId  driverId  constructorId number  grid position  \
0         1      18         1              1     22     1        1   
1         2      18         2              2      3     5        2   
2         3      18         3              3      7     7        3   
3         4      18         4              4      5    11        4   
4         5      18         5              1     23     3        5   

  positionText  positionOrder  points  laps         time milliseconds  \
0            1              1    10.0    58  1:34:50.616      5690616   
1            2              2     8.0    58       +5.478      5696094   
2            3              3     6.0    58       +8.163      5698779   
3            4              4     5.0    58      +17.181      5707797   
4            5              5     4.0    58      +18.014      5708630   

  fastestLap rank fastestLapTime fastestLapSpeed  statusId  
0         39    2       1:27.452         218.300         1  
1         41    3       1:27.739         217.586         1  
2         41    5       1:28.090         216.719         1  
3         58    7       1:28.603         215.464         1  
4         43    1       1:27.418         218.385         1

First off, comparing the number of raceIds in this dataset to that of df_lap_times it would seem that each driver has only one entry per race which is useful to know.

resultId I would guess is a reference for other datasets to use that we haven't seen come up yet so it will be dropped. constructorId is a reference for a dataset that hasn't been loaded as this is purely a driver analysis so it doesn't matter what team they belong to for my purposes. statusId is a reference to yet another database I haven't loaded and whose purpose I'm not sure of yet, I'll be dropping it for now.

In [30]:
df_results.drop(labels=['resultId', 'constructorId', 'statusId'], axis='columns', inplace=True)

In [31]:
df_results.head()

raceId  driverId number  grid position positionText  positionOrder  points  \
0      18         1     22     1        1            1              1    10.0   
1      18         2      3     5        2            2              2     8.0   
2      18         3      7     7        3            3              3     6.0   
3      18         4      5    11        4            4              4     5.0   
4      18         5     23     3        5            5              5     4.0   

   laps         time milliseconds fastestLap rank fastestLapTime  \
0    58  1:34:50.616      5690616         39    2       1:27.452   
1    58       +5.478      5696094         41    3       1:27.739   
2    58       +8.163      5698779         41    5       1:28.090   
3    58      +17.181      5707797         58    7       1:28.603   
4    58      +18.014      5708630         43    1       1:27.418   

  fastestLapSpeed  
0         218.300  
1         217.586  
2         216.719  
3         215.464  
4         218.385

#### Include Only Recent Years

Now, using the data we've loaded we can try and limit the size of the database by removing any entries that aren't from races in the years 2023-2024.

In [32]:
# this list can be modified to include any desired years
included_years = [2023, 2024]

In [33]:
df_lap_times.head()

raceId  driverId  lap  position      time  milliseconds
0     841        20    1         1  1:38.109         98109
1     841        20    2         1  1:33.006         93006
2     841        20    3         1  1:32.713         92713
3     841        20    4         1  1:32.803         92803
4     841        20    5         1  1:32.342         92342

In [34]:
# create new dataframe only including the selected years, reset index afterwards
df_races = df_races[df_races['year'].isin(included_years)].reset_index(drop=True)

# go through df_lap_times and df_results and filter out any entries whose raceIds dont' match the ones in df_races
included_raceIds = df_races['raceId']
df_lap_times = df_lap_times[df_lap_times['raceId'].isin(included_raceIds)].reset_index(drop=True)
df_results = df_results[df_results['raceId'].isin(included_raceIds)].reset_index(drop=True)

In [35]:
df_races.describe()

raceId         year      round  circuitId
count    46.000000    46.000000  46.000000  46.000000
mean   1121.391304  2023.521739  12.021739  33.173913
std      13.594898     0.505047   6.744509  29.461315
min    1098.000000  2023.000000   1.000000   1.000000
25%    1110.250000  2023.000000   6.250000   9.500000
50%    1121.500000  2024.000000  12.000000  19.500000
75%    1132.750000  2024.000000  17.750000  69.750000
max    1144.000000  2024.000000  24.000000  80.000000

In [36]:
df_lap_times.describe()

raceId      driverId           lap      position  milliseconds
count  36908.000000  36908.000000  36908.000000  36908.000000  3.690800e+04
mean    1115.017205    750.850493     30.832665      9.781592  9.710750e+04
std        9.896591    256.579438     18.473300      5.420859  1.033211e+05
min     1098.000000      1.000000      1.000000      1.000000  6.701200e+04
25%     1106.000000    822.000000     15.000000      5.000000  8.077675e+04
50%     1115.000000    840.000000     30.000000     10.000000  8.962700e+04
75%     1124.000000    848.000000     45.000000     14.000000  9.853425e+04
max     1131.000000    860.000000     78.000000     20.000000  2.625541e+06

In [37]:
df_results.describe()

raceId    driverId        grid  positionOrder      points  \
count   679.000000  679.000000  679.000000     679.000000  679.000000   
mean   1115.341679  754.821797    9.880707      10.486009    5.101620   
std      10.042442  251.560104    5.812627       5.763226    7.250469   
min    1098.000000    1.000000    0.000000       1.000000    0.000000   
25%    1107.000000  822.000000    5.000000       5.500000    0.000000   
50%    1115.000000  840.000000   10.000000      10.000000    1.000000   
75%    1124.000000  848.000000   15.000000      15.000000    9.500000   
max    1132.000000  860.000000   20.000000      20.000000   26.000000   

             laps  
count  679.000000  
mean    55.770250  
std     16.110247  
min      0.000000  
25%     51.000000  
50%     57.000000  
75%     68.000000  
max     78.000000

Min/Max Year for df_races looks right and it has a total of 46 raceIds.  
Min/Max raceId for df_results and df_lap_times look like they both span about 46 unqiue Ids! (that took an alarmingly small amount of time to get right)

#### Remove Inactive Drivers from df_drivers

In [38]:
df_drivers.describe()

driverId
count  859.000000
mean   430.059371
std    248.213115
min      1.000000
25%    215.500000
50%    430.000000
75%    644.500000
max    860.000000

In [39]:
df_drivers = df_drivers[df_drivers['driverId'].isin(df_results['driverId'])].reset_index(drop=True)
df_drivers.describe()

driverId
count   23.000000
mean   767.652174
std    241.919898
min      1.000000
25%    823.500000
50%    842.000000
75%    853.500000
max    860.000000

Evidently there are only 23 drivers left in the database which is around the number we want

#### Summary

There are now three processed databases to use:

##### df_drivers

Contains reference numbers and names for each driver

In [40]:
df_drivers.head()

driverId   driverRef number code
0         1    hamilton     44  HAM
1         4      alonso     14  ALO
2       842       gasly     10  GAS
3       807  hulkenberg     27  HUL
4       815       perez     11  PER

##### df_lap_times 

Contains information pertaining to every single lap each driver completed in each race, with lap time in min/sec/millisec format and also milliseconds.

In [41]:
df_lap_times.head()

raceId  driverId  lap  position      time  milliseconds
0    1098       830    1         1  1:39.019         99019
1    1098       830    2         1  1:37.974         97974
2    1098       830    3         1  1:38.006         98006
3    1098       830    4         1  1:37.976         97976
4    1098       830    5         1  1:38.035         98035

##### df_races

Contains references for decoding information in other databses, such as linking raceId to the track name and specific date and time the race occurred at.

In [42]:
df_races.head()

raceId  year  round  circuitId                      name        date  \
0    1098  2023      1          3        Bahrain Grand Prix  2023-03-05   
1    1099  2023      2         77  Saudi Arabian Grand Prix  2023-03-19   
2    1100  2023      3          1     Australian Grand Prix  2023-04-02   
3    1101  2023      4         73     Azerbaijan Grand Prix  2023-04-30   
4    1102  2023      5         79          Miami Grand Prix  2023-05-07   

       time  
0  15:00:00  
1  17:00:00  
2  05:00:00  
3  11:00:00  
4  19:30:00

##### df_results

Contains information pertaining to the results of the races themselves, including the starting positions of each driver, the finishing positions of each driver, the amount of time it took to complete the race, and more.

In [43]:
df_results.head()

raceId  driverId number  grid position positionText  positionOrder  points  \
0    1098       830      1     1        1            1              1    25.0   
1    1098       815     11     2        2            2              2    18.0   
2    1098         4     14     5        3            3              3    15.0   
3    1098       832     55     4        4            4              4    12.0   
4    1098         1     44     7        5            5              5    10.0   

   laps         time milliseconds fastestLap rank fastestLapTime  \
0    57  1:33:56.736      5636736         44    6       1:36.236   
1    57      +11.987      5648723         37    7       1:36.344   
2    57      +38.637      5675373         36    5       1:36.156   
3    57      +48.052      5684788         37   14       1:37.130   
4    57      +50.977      5687713         36   10       1:36.546   

  fastestLapSpeed  
0         202.452  
1         202.225  
2         202.620  
3         200.588  
4         201.802

### Data Visualization